Feature/falling layout

ibcylon opened the pull request:
## Motivation ⍰

- Moya Sample Data를 이용한 Stub 생성, request 시 sample Data로 Test 가능, comment 통해 설명드리겠습니다.
- [우아한형제들 참조](https://techblog.woowahan.com/2704/)


<br>

## Key Changes 🔑

- Moya Stub
- HeartList Profile Navigation
- chatRoom navigation

<br>

<img src="https://github.com/THT-Team/THT-iOS/assets/25360781/902c8f60-ef35-47e9-a3b3-41933b946562" width=25%>

## To Reviewers 🙏🏻

- 지금 시뮬레이터를 14 plus로 하는데 피그마 값을 쓰면 좀 어색하더라구요, 비율로 하긴 했는데 추후 피그마 값으로 변경하겠습니다.
- 지금 토픽 바가 1줄일 때랑 2줄일 때 높이 다르게 해야하는데 좋은 방법 아시면 알려주세요! 현재는 라인 수를 구해서 constraint를 remake하고 있습니다.
- 좋아요 리스트 -> 프로필 -> 대화힉 -> 채팅룸 시나리오 일 때. 프로필화면에서 대화하기를 누르면 프로필화면이 dismiss()되고 채팅룸으로 이동하게 만들어야 할 텐데, delegate말고, 다른 방식으로 구현해보신 적 있으신가요?

<br>

## Linked Issue 🔗

- #29
- #31 
- #28 
- #30 

## 해야할 일

- 다음에를 통한 셀 삭제 시 dataSource 변경
- 좋아요 페이지네이션 & refresh와 dataSource sync
- image 로드
- 읽은 좋아요 레드닷 local에 저장 북마크 표시 생각해보기

ibcylon committed:
Add: topicBar color

ibcylon committed:
Add: HeartAPI DTO

ibcylon committed:
Add: HeartTarget and API Sample Data for Stub

ibcylon committed:
Feat: Stub ProviderProtocol

ibcylon committed:
Feat: Heart List Layout

ibcylon committed:
Add: Design System Component

PaddigLabel, TopicBar, CollectionHeaderView

Add: Observerable Util for Void and Driver Empty

Add CollectionView util for generic reusable sugar

Add: calculate function for topic dynamic hegiht for lines

ibcylon committed:
Chore: RxDataSource Dependency

ibcylon committed:
Add: Profile Scroll View Layout

ibcylon committed:
Refact: singleton to heartService dependency

Minny27 commented on Falling/Sources/DataLayer/Network/ProviderProtocol.swift:
```diff
@@ -0,0 +1,55 @@
+
+extension ProviderProtocol {
+  static func consProvider(
+    _ isStub: Bool = false,
```
isStub를 파라미터의 참, 거짓에 따라서 샘플 응답과 실제 응답을 나눌 수 있군요
배워갑니다

Minny27 commented on Falling/Sources/Foundation/Component/TFTopicBarView.swift:
```diff
@@ -0,0 +1,104 @@
+    let label = UILabel()
+    label.textColor = FallingAsset.Color.neutral50.color
+    label.font = UIFont.thtSubTitle2Sb
+    label.numberOfLines = 2
```
최대 몇 줄 제한이 있는 건가요?

Minny27 committed:
[#26]Refactor: 토픽 바 컨텐츠 라벨 동적 높이 리팩토링

- 컨텐츠 라벨 라인 수를 0으로 설정
- 원리는 UIKit 뷰를 auto layout으로 설정하고 높이와 너비를 고정하지 않은 상태에서
intrinsic size로 내부 사이즈를 설정하는데, 해당 라벨에 텍스트가 할당되면(그 값이 변경되면)
invalidateIntrinsicContentSize()로 intrinsic content size에 맞게 제약조건을 다시 맞춤

Minny27 committed:
[#26]Refactor: 컴포넌트 뷰 TFBaseView 상속받기

- 기본적으로 구현해야하는 view의 형태를 base를 상속받아서 처리하기 위함
- subview를 추가할 때 subviews로 수정
- stackview도 확장으로 addArrangedSubviews 함수 구현

Minny27 commented on Falling/Sources/Util/UICollectionView+Utils.swift:
```diff
@@ -0,0 +1,35 @@
+public extension UICollectionView {
+
+  func register<T: UICollectionViewCell>(cellType: T.Type) {
+    self.register(cellType, forCellWithReuseIdentifier: T.reuseIdentifier)
```
이런식으로 만드는 건 생각 안 해봤었는데 좋은 거 같네요!

Minny27 commented:
※ 수정 사항
1. 토픽 뷰 라벨 라인 개수 0으로 설정
- 토픽 뷰 라벨 동적 높이 설정은 라인을 0으로 주면 동적으로 높이를 설정해줘서 동적 높이 설정으로 변경했습니다.
원리는 UIKit 뷰를 auto layout으로 설정하고 높이와 너비를 고정하지 않은 상태에서
intrinsic size로 내부 사이즈를 설정하는데, 해당 라벨에 텍스트가 할당되면(그 값이 변경되면)
invalidateIntrinsicContentSize()로 intrinsic content size에 맞게 제약조건을 다시 맞춘다고 하네요

2. 컴포넌트들도 base를 상속받기
- 부모의 틀을 유지하기 위함
- addsubviews, addArrangedSubviews 함수로 서브 뷰들 추가하도록 수정
